### PR TITLE
Add feature for restarting

### LINKED ISSUE #273: Improve features for restarting from checkpoints
In order to ensure we can rescue runs that crash in the middle of epochs we need to improve the quality of our checkpointing and restarting mechanisms
- See [here](https://discuss.huggingface.co/t/how-to-continue-training-and-not-overwrite-checkpoint-number/25370) for how models are saved when continuing (we’ll overwrite checkpoints). We already have something implemented but need to decide how we handle restarting in general e.g. need to consider learning rate scheduler, which data we have already trained on etc… if we do this.

Two options: use `from_pretrained` or `resume_from_checkpoint` in `Trainer.train()` maybe including a `uuid` for each run.
    - how we efficiently restart on a dataset?
    - how to save down dataloader state with HF?
    - consider [here](https://github.com/huggingface/transformers/blob/v4.27.2/src/transformers/trainer_pt_utils.py#L1022) and [here](https://huggingface.co/docs/accelerate/usage_guides/checkpoint) for more information

### PR DESCRIPTION
**Pull request by jackapbutler**
closes #273 
- default to restart when possible
- but allow customisation through configuration

### CODE REVIEW

File: experiments/scripts/run_tune.py [hunk 1]
```diff
@@ -32,6 +32,38 @@
...
+        - a value of False to not load a checkpoint
+    """
+    if isinstance(restart_checkpoint, str):
+        # if specific checkpoint provided
+        if os.path.isdir(restart_checkpoint):
+            return restart_checkpoint
+        else:
+            raise ValueError(f"checkpoint cannot be found at {restart_checkpoint}")
+
+    elif isinstance(restart_checkpoint, bool) and restart_checkpoint:
```
Comment (bethanyconnolly) [inline experiments/scripts/run_tune.py:52]: maybe this could be 
`elif restart_checkpoint is True:`
because restart_checkpoint can only be Union[str, bool] so if it isn't caught by the isinstance(restart_checkpoint, str) it would have to be a bool at this stage?

**Review by bethanyconnolly — APPROVED** [review]

File: experiments/scripts/run_tune.py (hunk 1 above)
Comment (jackapbutler) [reply experiments/scripts/run_tune.py:52]: yes agreed
